expint/polylog: unicode print subscripts in more cases

bsach64 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->
changed _print_expint and _print_polylog to help pretty print. 
Added _opportunistic_subscripter function for the same
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #25312 


#### Brief description of what is fixed or changed
Earlier printing expint & polylog functions with two symbols such as expint(s, z) did not subscript but now does as Eₛ(z)

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
  * Fixed a bug with printing of expint and polylog.
<!-- END RELEASE NOTES -->

bsach64 committed:
printing.pretty: prints expint, polylog with opportunistic subscripts
Earlier pretty printing did not work for
expint(s, z) but worked for E₃(z).
Added _opportunistic_subscripter function to pretty_symbology.py
It checks whether all characters of the argument are subscriptable.
addresses issue #25312
prints Eₛ(z) instead of expint(s, z)

bsach64 committed:
author: add Bhavik Sachdev to .mailmap

sylee957 commented on sympy/printing/pretty/pretty_symbology.py:
```diff
@@ -641,3 +641,6 @@ def line_width(line):
     return len(line.translate(_remove_combining))
+
+def _opportunistic_subscripter(e):
+    return all(character in sub for character in str(e.args[0]))
```
Why not get string from `self._print` than `str`.

bsach64 replied:
Done!

bsach64 committed:
Uses self._print instead of str to get the string

cbm755 commented on sympy/printing/pretty/pretty_symbology.py:
```diff
@@ -641,3 +641,6 @@ def line_width(line):
     return len(line.translate(_remove_combining))
+
+def _opportunistic_subscripter(subscript):
+    return all(character in sub for character in subscript)
```
Also maybe this shouldn't be private (?)  In which case, `is_subscriptable_in_unicode()` seems better.  It will then need a docstring.

bsach64 replied:
Done!

cbm755 commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1811,7 +1811,7 @@ def _print_DiracDelta(self, e):
     def _print_expint(self, e):
-        if e.args[0].is_Integer and self._use_unicode:
+        if _opportunistic_subscripter(self._print(e.args[0])) and self._use_unicode:
             return self._print_Function(Function('E_%s' % e.args[0])(e.args[1]))
```
Following up a bit on @sylee957's comment...  We need to be careful here: you can't ask permission about the `_print(args[0])` but then use the the non-printed `args[0]`.

Probably we can think of several fixes...

Alternatively: "zen of python" says "better to ask forgiveness than permission".  So another possibility is to code it in a way that this possibility is avoided?

bsach64 replied:
I also did write a fix for this ! (just forgot to comment it here)
Defined a variable called `subscript` that is used by both `is_subscriptable_in_unicode()` and `self._print_Function`

bsach64 committed:
Made Changes According to Reviews

Renamed _opportunistic_subscripter to is_subscriptable_in_unicode
Changed order of tests to slightly improve performance
Added docstring for is_subscriptable_in_unicode
Renamed test function for consistency with other functions
Uses _print(e.args[0]) instead of non-printed e.args[0]

bsach64 committed:
Merge branch 'sympy:master' into fix-sub-print

asmeurer commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -7508,7 +7507,7 @@ def test_print_polylog():
 
 
-@XFAIL  # Issue #25312
+# Issue #25312
 def test_print_expint_polylog_symbolic_order():
```
I much prefer this style to `test_issue_number`. I know we have a lot of tests named like that, but I'd much rather have a test function that specifies what it is testing than one that is named with a number.

bsach64 replied:
I changed it back to the way it was.

bsach64 committed:
Renamed test function to be more descriptive

bsach64 commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1664,7 +1664,7 @@ def _print_GeometryEntity(self, expr):
 
     def _print_polylog(self, e):
-        if e.args[0].is_Integer and self._use_unicode:
+        if _opportunistic_subscripter(self._print(e.args[0])) and self._use_unicode:
```
Should have caught that myself! Now fixed!

bsach64 committed:
author: added email

bsach64 commented:
Hey, just checking is there something more to be done? (Or something I did wrong)
